rules/TypeDeclaration/Rector/Class_/PropertyTypeFromStrictSetterGetterRector.php

```diff
@@ -0,0 +1,163 @@
+use Rector\PHPStanStaticTypeMapper\Enum\TypeKind;
+use Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer\GetterTypeDeclarationPropertyTypeInferer;
+use Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer\SetterTypeDeclarationPropertyTypeInferer;
+use Rector\VersionBonding\Contract\MinPhpVersionInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\TypeDeclaration\Rector\Class_\PropertyTypeFromStrictSetterGetterRector\PropertyTypeFromStrictSetterGetterRectorTest
+ */
+final class PropertyTypeFromStrictSetterGetterRector extends AbstractRector implements MinPhpVersionInterface
+{
+    public function __construct(
+        private readonly GetterTypeDeclarationPropertyTypeInferer $getterTypeDeclarationPropertyTypeInferer,
+        private readonly SetterTypeDeclarationPropertyTypeInferer $setterTypeDeclarationPropertyTypeInferer
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add property type based on strict setter and getter method', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    private $name = 'John';
+
+    public function setName(string $name): void
+    {
+        $this->name = $name;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+}
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    private string $name = 'John';
+
+    public function setName(string $name): void
+    {
+        $this->name = $name;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $hasChanged = false;
+
+        foreach ($node->getProperties() as $property) {
+            if ($property->type instanceof Node) {
+                continue;
+            }
+
+            if (! $property->isPrivate()) {
+                continue;
+            }
+
+            $getterSetterPropertyType = $this->matchGetterSetterIdenticalType($property, $node);
+            if (! $getterSetterPropertyType instanceof Type) {
+                continue;
+            }
+
+            if (! $this->isDefaultExprTypeCompatible($property, $getterSetterPropertyType)) {
+                continue;
+            }
+
+            $propertyTypeDclaration = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode(
+                $getterSetterPropertyType,
+                TypeKind::PROPERTY
+            );
+            if (! $propertyTypeDclaration instanceof Node) {
+                continue;
+            }
+
```

Comment:
Make property type guard is needed here

https://github.com/rectorphp/rector-src/blob/b98de3edc21312c2e5483b6a120ccb04642f5346/rules/TypeDeclaration/Rector/Property/TypedPropertyFromAssignsRector.php#L110

Reply:
I see. Could you handle this and add a test case too?

Reply:
I will try.

Reply:
here the PR https://github.com/rectorphp/rector-src/pull/3319